guides/internals.rst

```diff
@@ -0,0 +1,342 @@
+
+There are a number of cases where we find ourself with a user-provided function
+(where the "user" might still be something that is entirely our code) and we
+want to pass a whole bunch of different examples to it in order to achieve some
+result. Currently this includes each of the main engine, the Shrinker (in
+engine.py) and the minimizer, but there are likely to be more in future.
+
+We typically organise such things in terms of an object that you create with
+the function and possibly an initial argument that stores these on self and
+has some ``run`` or similar method. They then run for a while, repeatedly
+calling the function they were given.
+
+Generally speaking they do not call the function directly, but instead wrap
+calls to it. This allows them to implement a certain amount of decision caching,
+e.g. avoiding trying the same shrink twice, but also gives us a place where we
+can update metadata about the search process.
+
+For objects whose goal is some form of optimisation (Shrinker, Minimizer) one
+of the pieces of metadata they will typically track is a "current target". This
+is typically the best example they have seen so far. By wrapping every call to
+the predicate, we ensure that we never miss an example even when we're passing
+through other things.
+
+For objects whose goal is some broader form of search (currently only
+``ConjectureRunner``) this also allows them to keep track of *other* examples
+of interest. For example, as part of our multiple bug discovery,
+``ConjectureRunner`` keeps track of the smallest example of each distinct
+failure that it has seen, and updates this automatically each time the test
+function is called. This means that if during shrinking we "slip" and find a
+different bug than the one we started with, we will *not* shrink to that, but
+it will get remembered by the runner if it was either novel or better than our
+current example.
+
+~~~~~~~~~~~
+Weird Loops
+~~~~~~~~~~~
+
+The loops inside a lot of the engine look very strange and unidiomatic. For
+example:
+
+.. code-block:: python
+
+        i = 0
+        while i < len(self.intervals):
+            u, v = self.intervals[i]
+            if not self.incorporate_new_buffer(
+                self.shrink_target.buffer[:u] + self.shrink_target.buffer[v:]
+            ):
+                i += 1
+
+
+The more natural way to write this in Python would be:
+
+.. code-block:: python
+
+        for u, v in self.intervals:
+            self.incorporate_new_buffer(
+                self.shrink_target.buffer[:u] + self.shrink_target.buffer[v:]
+            )
+
+This way of writing the loop would be *entirely wrong*.
+
+Every time ``incorporate_new_buffer`` succeeds, it changes the shape of the
+current shrink target. This consequently changes the shape of intervals, both
+its particular values and its current length - on each loop iteration the loop
+might stop either because ``i`` increases or because ``len(self.intervals)``
+decreases.
```

Comment:
I'd add a note that starting a fresh loop each time `self.intervals` changes would have quadratic rather than linear complexity.